### PR TITLE
Add a Name field to a IR Route Destination

### PR DESCRIPTION
**Pull request by arkodg**
* Allows us to reduce and reuse the number of xds clusters created which are per xRoute per rule 
Fixes: https://github.com/envoyproxy/gateway/issues/551

* Removes the hostname from the xds cluster name since the xds cluster name is no longer tied to the ir route name
Fixes: https://github.com/envoyproxy/gateway/issues/1785

* Also append `Kind` (`httproute` / `grpcroute`) to the IR Route and IR RouteDestination names so the generated xds routes and clusters are unique across different GAPI resource types and dont collide

### CODE REVIEW

**Comment by arkodg** [comment]
looks like I broke the Redirect conformance tests
```
2023-08-16T22:10:53.4357382Z     --- FAIL: TestGatewayAPIConformance/HTTPRouteRedirectHostAndStatus (60.04s)
2023-08-16T22:10:53.4358026Z     --- FAIL: TestGatewayAPIConformance/HTTPRouteRedirectPath (60.02s)
2023-08-16T22:10:53.4358767Z     --- FAIL: TestGatewayAPIConformance/HTTPRouteRedirectPortAndScheme (60.08s)
2023-08-16T22:10:53.4361651Z     --- FAIL: TestGatewayAPIConformance/HTTPRouteRedirectPort (60.02s)
2023-08-16T22:10:53.4362251Z     --- FAIL: TestGatewayAPIConformance/HTTPRouteRedirectScheme (60.02s)
```

looking into this rn

File: internal/xds/translator/cluster.go [hunk 1]
```diff
@@ -69,28 +69,28 @@ func buildXdsCluster(routeName string, tSocket *corev3.TransportSocket, protocol
 	return cluster
```
Comment (zhaohuabing) [inline internal/xds/translator/cluster.go:69]: Maybe also change the parameter `routeName` to `clusterName`?

```
func buildXdsCluster(routeName string, tSocket *corev3.TransportSocket, protocol ProtocolType, endpointType EndpointType) 
```

**Review by zhaohuabing — COMMENTED** [review]
Please correct me if I am wrong: 

This PR creates one cluster per route rule, which removes the duplicated clusters when a route is attached to multiple gateway listeners. But if a backend service is used in multiple routes, there still are duplicated clusters for that service. Should we consider creating one cluster per Backend service?

**Comment by arkodg** [comment]
> Please correct me if I am wrong:
> 
> This PR creates one cluster per route rule, eliminating the duplicated clusters when a route is attached to multiple gateway listeners. However, if a backend service is referenced in more than one route, there still are duplicated clusters created for that service. Should we consider creating one cluster per backend service?

good point, here's the thing, the xds cluster also contains other attributes such as loadbalancing and resiliency (outlier detection) that may be applied at a per xRoute per Rule level via a PolicyAttachment, so we need to create different clusters for those cases :)

**Review by zhaohuabing — APPROVED** [review]

**Review by AliceProxy — APPROVED** [review]
Looks good to me and the xDS names are clean

**Review by zirain — DISMISSED** [review]

**Comment by arkodg** [comment]
lets wait another day for https://github.com/envoyproxy/gateway/pull/1702 to go in, before this PR goes in, 1702 has already had to rebase a few times

**Review by zirain — APPROVED** [review]

**Pull request merged**